MSFT_O365User: Creating new user fails when "LicenseAssignment" array is empty

<!--
    Your feedback and support is greatly appreciated, thanks for contributing!

    ISSUE TITLE:
    Please prefix the issue title with the resource name, e.g.
    'ResourceName: Short description of my issue'

    ISSUE DESCRIPTION (this template):
    Please provide information regarding your issue under each header below.
    Write N/A under any headers that do not apply to your issue, or if the
    information is not available.

    NOTE! Sensitive information should be obfuscated.

    PLEASE KEEP THE HEADERS.

    You may remove this comment block, and the other comment blocks,
    but please keep the headers.
-->
#### Details of the scenario you tried and the problem that is occurring
We are using Microsoft365DSC within Azure DevOps Pipelines to make sure certain break-glass accounts for our tenant are never changed or deleted. These accounts don't need any licenses by default - they'll be assigned only during the time they'll actually be used.

These accounts are being created and maintained using the MSFT_O365User resource. However, when adding a new account in the config section for that module (so the resource will create a new account), while also setting the property "LicenseAssignment" to an empty array, Set-TargetResource will fail with an error for that account. The accompanying log messages are very confusing, since they will state that 1 new license is to be assigned (should be 0) and then saying that the license "{}" couldn't be assigned.
After checking the code at least for the logs showing "{}" instead of the license name I know the root cause: the log message includes the variable "$newLicenseAssignment" which was never initialized before.

The actual outcome of the whole process is correct, though: The new user account was created and has no licenses assigned. And yet our pipeline stage for applying this configuration is always coming up as failed due to that error.
#### Verbose logs showing the problem
```
2020-08-28T12:26:39.2298315Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777] Setting configuration 
2020-08-28T12:26:39.2306459Z of Office 365 User XXXXX@XXXXX.de
2020-08-28T12:26:41.6263603Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777] Getting configuration 
2020-08-28T12:26:41.6269330Z of Office 365 User XXXXX@XXXXX.de
2020-08-28T12:26:42.3324078Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777] Getting Office 365 
2020-08-28T12:26:42.3329518Z User XXXXX@XXXXX.de
2020-08-28T12:26:42.4305997Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777] The specified User 
2020-08-28T12:26:42.4311354Z {XXXXX@XXXXX.de} doesn't already exist.
2020-08-28T12:26:42.4937892Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777] Creating Office 365 
2020-08-28T12:26:42.4943402Z User XXXXX@XXXXX.de
2020-08-28T12:26:43.2321276Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777] Assigning 1 licences 
2020-08-28T12:26:43.2327324Z to new user
2020-08-28T12:26:43.3597286Z VERBOSE: [fv-az740]:                            [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777]  Logging a new Error
2020-08-28T12:26:43.5616086Z VERBOSE: [fv-az740]: LCM:  [ End    Set      ]  [[O365User]9a9c201c-b8ee-4ce7-8726-1d66bc4b8777]  in 4.3870 seconds.
2020-08-28T12:26:43.6591864Z ##[error]PowerShell DSC resource MSFT_O365User  failed to execute Set-TargetResource functionality with error message: Could 
not assign license {} to user {XXXXX@XXXXX.de} 

    + CategoryInfo          : InvalidOperation: (:) [], CimException

    + FullyQualifiedErrorId : ProviderOperationExecutionFailure

    + PSComputerName        : localhost
```
#### Suggested solution to the issue

- Change the logic for assigning licenses within Set-TargetResource of the MSFT_O365User resource so it skips license assignments if "LicenseAssignment" is an empty array or $null in the corresponding config section for users which are being newly created.

- Update the error message to use the correct variable.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
O365User 9a9c201c-b8ee-4ce7-8726-1d66bc4b8777
        {
            DisplayName          = "XXXX";
            Ensure               = "Present";
            FirstName            = "XXXX";
            GlobalAdminAccount   = $Credsglobaladmin;
            LastName             = "XXXX";
            LicenseAssignment    = @();
            Password             = $Credsglobaladmin;
            PasswordNeverExpires = $true;
            UsageLocation        = "DE";
            UserPrincipalName    = "XXXX@XXXX.de";
            UserType             = "Member";
        }
```

#### The operating system the target node is running
<!--
    Please provide as much as possible about the target node, for example
    edition, version, build and language.
    On OS with WMF 5.1 the following command can help get this information.

    Get-ComputerInfo -Property @(
        'OsName',
        'OsOperatingSystemSKU',
        'OSArchitecture',
        'WindowsVersion',
        'WindowsBuildLabEx',
        'OsLanguage',
        'OsMuiLanguages')
-->
Azure DevOps Pipeline Agent with specification "vs2017-win2016"
#### Version of the DSC module that was used ('dev' if using current dev branch)
1.20.805.1
(We can't install the Dev version on the agent due to already loaded modules preventing the installation of that)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,12 @@
 
 ## UNRELEASED
 
+* O365User
+  * Fixed an issue where we were trying to assign an empty
+    license to a user if an empty array was passed for
+    LicenseAssignment.
+* SCComplianceSearchAction
+  * Added 'Preview' as a supported value for Action;
 * ReverseDSC
   * Fixed an issue where a newline was missing in the
     credentials section when a certificate password was

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_O365User/MSFT_O365User.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_O365User/MSFT_O365User.psm1
@@ -446,10 +446,14 @@ function Set-TargetResource
 
         try
         {
-            Write-Verbose -Message "Assigning $($licenses.Count) licences to new user"
-            Set-AzureADUserLicense -ObjectId $UserPrincipalName `
-                -AssignedLicenses $licenses `
-                -ErrorAction SilentlyContinue
+            if ($null -ne $licenses -and `
+            ($licenses.AddLicenses.Length -gt 0 -or $licenses.RemoveLicenses.Length -gt 0))
+            {
+                Write-Verbose -Message "Assigning $($licenses.Count) licences to new user"
+                Set-AzureADUserLicense -ObjectId $UserPrincipalName `
+                    -AssignedLicenses $licenses `
+                    -ErrorAction SilentlyContinue
+            }
         }
         catch
         {

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_SCComplianceSearchAction/MSFT_SCComplianceSearchAction.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_SCComplianceSearchAction/MSFT_SCComplianceSearchAction.psm1
@@ -6,7 +6,7 @@ function Get-TargetResource
     (
         [Parameter(Mandatory = $true)]
         [System.String]
-        [ValidateSet('Export', 'Purge', 'Retention')]
+        [ValidateSet('Export', 'Preview', 'Purge', 'Retention')]
         $Action,
 
         [Parameter(Mandatory = $true)]
@@ -148,7 +148,7 @@ function Set-TargetResource
     (
         [Parameter(Mandatory = $true)]
         [System.String]
-        [ValidateSet('Export', 'Purge', 'Retention')]
+        [ValidateSet('Export', 'Preview', 'Purge', 'Retention')]
         $Action,
 
         [Parameter(Mandatory = $true)]
@@ -301,7 +301,7 @@ function Test-TargetResource
     (
         [Parameter(Mandatory = $true)]
         [System.String]
-        [ValidateSet('Export', 'Purge', 'Retention')]
+        [ValidateSet('Export', 'Preview', 'Purge', 'Retention')]
         $Action,
 
         [Parameter(Mandatory = $true)]

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_SCComplianceSearchAction/MSFT_SCComplianceSearchAction.schema.mof
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_SCComplianceSearchAction/MSFT_SCComplianceSearchAction.schema.mof
@@ -1,7 +1,7 @@
 [ClassVersion("1.0.0.0"), FriendlyName("SCComplianceSearchAction")]
 class MSFT_SCComplianceSearchAction : OMI_BaseResource
 {
-    [Key, Description("The Action parameter specifies what type of action to define. Accepted values are Export, Retention and Purge."), ValueMap{"Export","Purge","Retention"}, Values{"Export","Purge","Retention"}] String Action;
+    [Key, Description("The Action parameter specifies what type of action to define. Accepted values are Export, Retention and Purge."), ValueMap{"Export","Preview","Purge","Retention"}, Values{"Export","Preview","Purge","Retention"}] String Action;
     [Key, Description("The SearchName parameter specifies the name of the existing content search to associate with the content search action. You can specify multiple content searches separated by commas.")] String SearchName;
     [Write, Description("The FileTypeExclusionsForUnindexedItems specifies the file types to exclude because they can't be indexed. You can specify multiple values separated by commas.")] String FileTypeExclusionsForUnindexedItems[];
     [Write, Description("The EnableDedupe parameter eliminates duplication of messages when you export content search results.")] Boolean EnableDedupe;

